### PR TITLE
Parameterize new overmap changes and adjust forests in the process

### PR DESCRIPTION
**Pull request by I-am-Erk**
#### Summary
Infrastructure "Parameterize new geography changes to overmaps"

#### Purpose of change

In #69912 I made a quick, dirty, and effective change to make cities increase in the East and South, while forests increase in the North and West. This should be something we can adjust and control with JSON settings

#### Describe the solution
Allow this to be adjusted and controlled with JSON settings.
- you can set a direction - N,E,S,W - in which forests or cities should increase. For cities, the increase will falloff in the adjacent cardinal direction, so if you set cities to only increase to the S, then if you travel SE or SW the megacity will fall away.
- You can set max values for these changes as well.
- Implicit to the increasing forest comes a decrease in city size and increase in spacing, so that we don't see a bunch of cities crammed into the tiny remaining open space in the forests.

This creates two new variables in overmap - urbanity and forestosity. These will be used for other things soon.

#### Testing

So much. Argh.
image
Here's a nice example of what we can see in the new forests. A tiny podunk little town surrounded by farms, orchards, and a mine.
image
here's a larger zoom out on four overmaps near that previous shot. I can just see the people of that small town, Lebanon, telling you you're gonna have to head up to Brandon, just past Newfields, if you want to get supplies.

The cities haven't changed substantially.

#### Additional context
Gosh, we need more roads now hey?

### CODE REVIEW

File: src/overmap.h [hunk 1]
```diff
@@ -200,6 +200,9 @@ class overmap
         const point_abs_om &pos() const {
             return loc;
         }
+        int get_urbanity() {
```
Comment (I-am-Erk) [inline src/overmap.h:203]:
```suggestion
        int get_urbanity() const {
```

**Comment by PatrikLundell** [comment]
We certainly need roads connecting the few isolated places. Although most of the added roads would probably pass through area monster territory, given how far you'd have to travel along them to get to the next location.

However, with your two changes, I'd have urbanity increase eastward (as you've already done), and arboreal abundance northward to create a variety grid.

**Review by ehughsbaird — COMMENTED** [review]
Given all the rest of the forest settings are in regional_settings, I think these tweaks to it would fit better there too, but the configuration for overmap generation seems to be split between those two places, so who knows :)
Also, I don't know that the game options support points, and I think the offset is a nice tweak.
<details>

<summary>Mega diff to move those to regional settings</summary>

```diff
diff --git a/data/core/game_balance.json b/data/core/game_balance.json
index ba76a2f980..b45f87add7 100644
--- a/data/core/game_balance.json
+++ b/data/core/game_balance.json
@@ -265,41 +265,6 @@
     "stype": "bool",
     "value": true
   },
-  {
-    "type": "EXTERNAL_OPTION",
-    "name": "OVERMAP_FOREST_INCREASE_NORTH",
-    "info": "Rate at which forest coverage of the map increases to the North, based on regional map settings thresholds.  0=no increase.",
-    "stype": "float",
-    "value": 0.04
-  },
-  {
-    "type": "EXTERNAL_OPTION",
-    "name": "OVERMAP_FOREST_INCREASE_EAST",
-    "info": "Rate at which forest coverage of the map increases to the East.  0=no increase.",
-    "stype": "float",
-    "value": 0
-  },
-  {
-    "type": "EXTERNAL_OPTION",
-    "name": "OVERMAP_FOREST_INCREASE_WEST",
-    "info": "Rate at which forest coverage of the map increases to the West.  0=no increase.",
-    "stype": "float",
-    "value": 0.02
-  },
-  {
-    "type": "EXTERNAL_OPTION",
-    "name": "OVERMAP_FOREST_INCREASE_SOUTH",
-    "info": "Rate at which forest coverage of the map increases to the South.  0=no increase.",
-    "stype": "float",
-    "value": 0
-  },
-  {
-    "type": "EXTERNAL_OPTION",
-    "name": "OVERMAP_FOREST_LIMIT",
-    "info": "Caps how high the forest threshold can rise.  No cities form at values over 0.4",
-    "stype": "float",
-    "value": 0.395
-  },
   {
     "type": "EXTERNAL_OPTION",
     "name": "OVERMAP_PLACE_SWAMPS",
diff --git a/data/json/regional_map_settings.json b/data/json/regional_map_settings.json
index 06cba104c0..2fa92aa5a0 100644
--- a/data/json/regional_map_settings.json
+++ b/data/json/regional_map_settings.json
@@ -361,6 +361,10 @@
       "noise_threshold_forest_thick": 0.25,
       "noise_threshold_swamp_adjacent_water": 0.3,
       "noise_threshold_swamp_isolated": 0.6,
+      "coverage_increase_limit": 0.395,
+      "coverage_increase_west_east": [ 0.02, 0 ],
+      "coverage_increase_south_north": [ 0, 0.04 ],
+      "coverage_increase_start_offset": [ 0, 0 ],
       "river_floodplain_buffer_distance_min": 3,
       "river_floodplain_buffer_distance_max": 15
     },
diff --git a/data/json/test_regions.json b/data/json/test_regions.json
index 56a947cd29..f4d8638ffd 100644
--- a/data/json/test_regions.json
+++ b/data/json/test_regions.json
@@ -55,6 +55,10 @@
       "noise_threshold_forest_thick": 0.7,
       "noise_threshold_swamp_adjacent_water": 0.4,
       "noise_threshold_swamp_isolated": 0.7,
+      "coverage_increase_limit": 0.395,
+      "coverage_increase_west_east": [ 0.02, 0 ],
+      "coverage_increase_south_north": [ 0, 0.04 ],
+      "coverage_increase_start_offset": [ 0, 0 ],
       "river_floodplain_buffer_distance_min": 1,
       "river_floodplain_buffer_distance_max": 6
     },
diff --git a/doc/REGION_SETTINGS.md b/doc/REGION_SETTINGS.md
index 179e89a6cf..386053f1bf 100644
--- a/doc/REGION_SETTINGS.md
+++ b/doc/REGION_SETTINGS.md
@@ -172,6 +172,10 @@ features are interpreted.
 | `noise_threshold_forest_thick`         | [0, 1], x > value spawns `forest_thick`.                               |
 | `noise_threshold_swamp_adjacent_water` | [0, 1], x > value spawns `forest_water` if forest near a waterbody.    |
 | `noise_threshold_swamp_isolated`       | [0, 1], x > value spawns `forest_water` if forest isolated from water. |
+| `coverage_increase_limit` | [0, 1], maximum value that forest noise threshold can decrease by. Values above 0.4 result in no cities generating. |
+| `coverage_increase_west_east` | Effective decrease in noise threshold for each overmap west/east of the offset point. |
+| `coverage_increase_south_north` | Effective decrease in noise threshold for each overmap south/north of the offset point. |
+| `coverage_increase_start_offset` | Overmap coordinates for which the values provided are used unmodified. Overmaps east/west or north/south will have these values adjusted by the coverage increase values. |
 | `river_floodplain_buffer_distance_min` | Minimum buffer distance in overmap terrains for river floodplains.     |
 | `river_floodplain_buffer_distance_max` | Maximum buffer distance in overmap terrains for river floodplains.     |
 
@@ -184,6 +188,10 @@ features are interpreted.
                "noise_threshold_forest_thick": 0.3,
                "noise_threshold_swamp_adjacent_water": 0.3,
                "noise_threshold_swamp_isolated": 0.6,
+               "coverage_increase_limit": 0.395,
+               "coverage_increase_west_east": [ 0.02, 0 ],
+               "coverage_increase_south_north": [ 0, 0.04 ],
+               "coverage_increase_start_offset": [ 0, 0 ],
                "river_floodplain_buffer_distance_min": 3,
                "river_floodplain_buffer_distance_max": 15
        }
diff --git a/src/overmap.cpp b/src/overmap.cpp
index 90f9b7c991..bfb603b013 100644
--- a/src/overmap.cpp
+++ b/src/overmap.cpp
@@ -5106,11 +5106,12 @@ void overmap::place_river( const point_om_omt &pa, const point_om_omt &pb )
 
 void overmap::calculate_forestosity()
 {
-    float northern_forest_increase = get_option<float>( "OVERMAP_FOREST_INCREASE_NORTH" );
-    float eastern_forest_increase = get_option<float>( "OVERMAP_FOREST_INCREASE_EAST" );
-    float western_forest_increase = get_option<float>( "OVERMAP_FOREST_INCREASE_WEST" );
-    float southern_forest_increase = get_option<float>( "OVERMAP_FOREST_INCREASE_SOUTH" );
-    const point_abs_om this_om = pos();
+    float northern_forest_increase = settings->overmap_forest.coverage_increase_south_north.second;
+    float eastern_forest_increase = settings->overmap_forest.coverage_increase_west_east.second;
+    float western_forest_increase = settings->overmap_forest.coverage_increase_west_east.first;
+    float southern_forest_increase = settings->overmap_forest.coverage_increase_south_north.first;
+    const point_abs_om this_om = pos() - settings->overmap_forest.coverage_increase_start_offset;
+
     if( western_forest_increase != 0 && this_om.x() < 0 ) {
         forest_size_adjust -= this_om.x() * western_forest_increase;
     }
@@ -5127,7 +5128,7 @@ void overmap::calculate_forestosity()
     //debugmsg( "forestosity = %1.2f at OM %i, %i", forestosity, this_om.x(), this_om.y() );
     // make sure forest size never totally overwhelms the map
     forest_size_adjust = std::min( forest_size_adjust,
-                                   get_option<float>( "OVERMAP_FOREST_LIMIT" ) - static_cast<float>
+                                   settings->overmap_forest.coverage_increase_limit - static_cast<float>
                                    ( settings->overmap_forest.noise_threshold_forest ) );
 }
 
diff --git a/src/regional_settings.cpp b/src/regional_settings.cpp
index 027cacd088..a420e1a73f 100644
--- a/src/regional_settings.cpp
+++ b/src/regional_settings.cpp
@@ -323,6 +323,16 @@ static void load_overmap_forest_settings(
                                     overmap_forest_settings.river_floodplain_buffer_distance_min, !overlay );
         read_and_set_or_throw<int>( overmap_forest_settings_jo, "river_floodplain_buffer_distance_max",
                                     overmap_forest_settings.river_floodplain_buffer_distance_max, !overlay );
+
+        read_and_set_or_throw<float>( overmap_forest_settings_jo, "coverage_increase_limit",
+                                      overmap_forest_settings.coverage_increase_limit, !overlay );
+
+        read_and_set_or_throw<std::pair<float, float>>( overmap_forest_settings_jo,
+                "coverage_increase_west_east", overmap_forest_settings.coverage_increase_west_east, !overlay );
+        read_and_set_or_throw<std::pair<float, float>>( overmap_forest_settings_jo,
+                "coverage_increase_south_north", overmap_forest_settings.coverage_increase_south_north, !overlay );
+        read_and_set_or_throw<point>( overmap_forest_settings_jo, "coverage_increase_start_offset",
+                                      overmap_forest_settings.coverage_increase_start_offset, !overlay );
     }
 }
 
diff --git a/src/regional_settings.h b/src/regional_settings.h
index 96ff0edf6a..427df3bede 100644
--- a/src/regional_settings.h
+++ b/src/regional_settings.h
@@ -196,6 +196,11 @@ struct overmap_forest_settings {
     int river_floodplain_buffer_distance_min = 3;
     int river_floodplain_buffer_distance_max = 15;
 
+    float coverage_increase_limit;
+    std::pair<float, float> coverage_increase_west_east;
+    std::pair<float, float> coverage_increase_south_north;
+    point coverage_increase_start_offset;
+
     overmap_forest_settings() = default;
 };
 
```

</details>

File: src/overmap.h [hunk 2]
```diff
@@ -445,6 +448,15 @@ class overmap
         // City Building
         overmap_special_id pick_random_building_to_place( int town_dist ) const;
 
+        // urbanity and forestosity are biome stats that can be used to trigger changes in biome.
+        int urbanity = 0;
```
[the remark this answers is not in the document]
Comment (I-am-Erk) [reply src/overmap.h:452]: While they're not needed right now, I'll be using them for some other things shortly and don't want to make any calls like that until it's certain.

**Comment by I-am-Erk** [comment]
> Given all the rest of the forest settings are in regional_settings, I think these tweaks to it would fit better there too, but the configuration for overmap generation seems to be split between those two places, so who knows :) Also, I don't know that the game options support points, and I think the offset is a nice tweak.

That's a really good change, but i'm a bit hesitant for two reasons: one, if I put it in by cp it won't get your credit, which is deserved, and two, I do plan on using different regional settings JSON based on the results of the new biome variables; that's not a hard barrier to storing this in regional settings, but it could get really weird if the stuff used to calculate them changes. That's not a hard barrier though, it can just be a 'user beware' kind of thing.

**Comment by I-am-Erk** [comment]
After reviewing your suggestions and working out how I want to do varied region settings, I think you're right and I can integrate it. I've added the suggestion you made but if you want to keep credit beyond mentioning you in commits, I could take it out and you could PR it in followup

**Comment by I-am-Erk** [comment]
Thanks mal

**Pull request merged**